lib/semantics/semantics.cpp

```diff
@@ -111,18 +111,20 @@ template<typename... C> class SemanticsVisitor : public virtual C... {
 
 using StatementSemanticsPass1 = ExprChecker;
 using StatementSemanticsPass2 = SemanticsVisitor<  //
-    AllocateChecker, ArithmeticIfStmtChecker, AssignmentChecker, CoarrayChecker,
-    DeallocateChecker, DoForallChecker, IfStmtChecker, IoChecker,
-    NullifyChecker, OmpStructureChecker, PurityChecker, ReturnStmtChecker,
-    StopChecker>;
+    AllocateChecker, ArithmeticIfStmtChecker, CoarrayChecker, DeallocateChecker,
+    DoForallChecker, IfStmtChecker, IoChecker, NullifyChecker,
+    OmpStructureChecker, PurityChecker, ReturnStmtChecker, StopChecker>;
+using StatementSemanticsPass3 = SemanticsVisitor<AssignmentChecker>;
```

Comment:
Adding a third pass over the parse tree seems expensive if we're just getting warning messages for obsolescent `FORALL` statements in return (is that the case?).  Can we avoid a third pass, or perhaps instantiate a localized one that only operates as a second pass over `FORALL` bodies when they appear (which is rare)?

Reply:
My assumption is that we will eventually need a third pass anyway as more checkers are added. If that's not the case, I agree this is overkill just for these FORALL checks.

In this case the conflict is only because I want to know when an assignment or concurrent-header is inside a FORALL vs. a DO CONCURRENT. SemanticsContext maintains a stack of constructs we are currently nested inside. That would be good enough except for the case of a FORALL statement, which wouldn't appear in the stack. We could get it to work by treating a FORALL statement like a construct (we would probably want to do the same with IF statements). Or perhaps rewrite FORALL statements into FORALL constructs early on so we only have that case to worry about.

Reply:
However this works out, it would be counterintuitive if non-construct `IF`, `FORALL`, &/or `WHERE` statements were not represented in the context used for semantic analysis of their single controlled statements.

Reply:
I've moved the check into check-do-forall.cpp where we were already collect FORALL index variables anyway. This allows us to get by with only two semantics passes.